src/util/color_spaces.ts

```diff
@@ -27,111 +41,101 @@ const Xn = 0.950470, // D65 standard referent
     y = Yn * lab2xyz(y);
     x = Xn * lab2xyz(x);
     z = Zn * lab2xyz(z);
-    return new Color(
+
+    return [
         xyz2rgb(3.2404542 * x - 1.5371385 * y - 0.4985314 * z), // D65 -> sRGB
         xyz2rgb(-0.9692660 * x + 1.8760108 * y + 0.0415560 * z),
         xyz2rgb(0.0556434 * x - 0.2040259 * y + 1.0572252 * z),
-        labColor.alpha
-    );
+        alpha,
+    ];
 }
 
-function interpolateLab(from: LABColor, to: LABColor, t: number) {
-    return {
-        l: interpolateNumber(from.l, to.l, t),
-        a: interpolateNumber(from.a, to.a, t),
-        b: interpolateNumber(from.b, to.b, t),
-        alpha: interpolateNumber(from.alpha, to.alpha, t)
-    };
+function xyz2rgb(x: number): number {
+    x = (x <= 0.00304) ? 12.92 * x : 1.055 * Math.pow(x, 1 / 2.4) - 0.055;
+    return (x < 0) ? 0 : (x > 1) ? 1 : x; // clip to 0..1 range
 }
 
-// HCL
-function rgbToHcl(rgbColor: Color): HCLColor {
-    const {l, a, b} = rgbToLab(rgbColor);
-    const h = Math.atan2(b, a) * rad2deg;
-    return {
-        h: h < 0 ? h + 360 : h,
-        c: Math.sqrt(a * a + b * b),
-        l,
-        alpha: rgbColor.a
-    };
+function lab2xyz(t: number): number {
+    return (t > t1) ? t * t * t : t2 * (t - t0);
 }
 
-function hclToRgb(hclColor: HCLColor): Color {
-    const h = hclColor.h * deg2rad,
-        c = hclColor.c,
-        l = hclColor.l;
-    return labToRgb({
-        l,
-        a: Math.cos(h) * c,
-        b: Math.sin(h) * c,
-        alpha: hclColor.alpha
-    });
+// sRGB -> HCL
+function rgbToHcl(rgbColor: RGBColor): HCLColor {
+    const [l, a, b, alpha] = rgbToLab(rgbColor);
+    const c = Math.sqrt(a * a + b * b);
+    const h = Math.round(c * 10000) ? constrainAngle(Math.atan2(b, a) * rad2deg) : NaN;
+    return [h, c, l, alpha];
 }
 
-function interpolateHue(a: number, b: number, t: number) {
-    const d = b - a;
-    return a + t * (d > 180 || d < -180 ? d - 360 * Math.round(d / 360) : d);
+// HCL -> sRGB
+function hclToRgb([h, c, l, alpha]: HCLColor): RGBColor {
+    h = isNaN(h) ? 0 : h * deg2rad;
+    return labToRgb([l, Math.cos(h) * c, Math.sin(h) * c, alpha]);
 }
 
-function interpolateHcl(from: HCLColor, to: HCLColor, t: number) {
-    return {
-        h: interpolateHue(from.h, to.h, t),
-        c: interpolateNumber(from.c, to.c, t),
-        l: interpolateNumber(from.l, to.l, t),
-        alpha: interpolateNumber(from.alpha, to.alpha, t)
-    };
+// HSL -> sRGB https://drafts.csswg.org/css-color-4/#hsl-to-rgb
+function hslToRgb([h, s, l, alpha]: HSLColor): RGBColor {
+    h = constrainAngle(h);
+    s /= 100;
+    l /= 100;
+
+    function f(n) {
+        const k = (n + h / 30) % 12;
+        const a = s * Math.min(l, 1 - l);
+        return l - a * Math.max(-1, Math.min(k - 3, 9 - k, 1));
+    }
+
+    return [f(0), f(8), f(4), alpha];
 }
 
 export const lab = {
-    forward: rgbToLab,
-    reverse: labToRgb,
-    interpolate: interpolateLab
+    fromRgb: rgbToLab,
```

Comment:
why not simply export the methods in this file instead of these objects?

Reply:
They neatly group functions related to a single color space.

Reply:
Please simply rename the methods to `hclToRgb` and the like and export those, the extra object creates confusion and looks as if there's an actual object that has a purpose in other files, which it doesn't.

Reply:
Done